blog: multi-user MCP with per-backend header forwarding (#2329)
envoyproxy/ai-gateway: site/blog/2026/2026-07-05-multi-user-mcp-header-forwarding.md

---
slug: multi-user-mcp-header-forwarding
title: "Multi-User MCP in Production: Per-User Identity at the Gateway Layer"
authors: [mohitgurnani]
tags: [features]

> description: How Envoy AI Gateway's per-backend header forwarding brings per-user identity to MCP tool calls — no OAuth infrastructure required — and when to reach for OAuth or token exchange instead.

Reviewer: ```suggestion
description: How Envoy AI Gateway's header forwarding gives MCP tool calls per-user identity without OAuth infrastructure — and when you need OAuth instead.
```

- **Token exchange (RFC 8693).** The user's corporate token means nothing to an external SaaS on a different identity provider. A token service exchanges it for a downstream token that still represents the same user. Right for per-user identity _across_ an IdP boundary — at the cost of real infrastructure.

There is also the full OAuth authorization code flow with browser consent, which the MCP specification mandates for interactive clients talking to external servers. It is the right answer when a human is present — and architecturally impossible for a headless agent firing at 3am with no browser.

| Pattern                   | Headless? | Per-user RBAC?   | New infrastructure? | Right when                                 |
| ------------------------- | --------- | ---------------- | ------------------- | ------------------------------------------ |
| OAuth auth code + PKCE    | ❌        | ✅               | OAuth AS per tool   | Interactive clients, external SaaS         |
| Client credentials        | ✅        | Fixed scope only | App registration    | Agent acts as itself                       |

> | Token exchange (RFC 8693) | ✅        | ✅               | STS + trust config  | Cross-IdP boundary                         |

Reviewer: Minor clarity nit: this table uses "STS" and "OAuth AS" without expanding them on first use, whereas "PAT" is spelled out earlier in the post ("personal access token (PAT)"). Might be worth spelling these out too (or a footnote) for readers less familiar with the acronyms.

![Decision path for choosing a downstream identity pattern](/img/blog/multi-user-mcp-decision.png)

## How Header Forwarding Works

The key design property is what _doesn't_ happen: Envoy AI Gateway's MCP proxy builds a fresh HTTP request for every upstream call. Client headers do not propagate to backend MCP servers by default — at all. Forwarding is an explicit, per-backend opt-in via `forwardHeaders` on each backend reference.

![Fan-out with per-backend header scoping](/img/blog/multi-user-mcp-fanout.png)

This matters most during fan-out. A single `tools/list` call fans out to every backend in the route, but each backend receives only the headers it explicitly opted into. A per-user Atlassian token configured for the Atlassian backend is never seen by the nine other backends in the same route. The absence of `forwardHeaders` on a backend means no identity propagation to that service, full stop.

Each entry can also rename the header on the way through: `name` selects the inbound header, and an optional `backendHeader` sets the name used toward the backend — useful when a backend expects a vendor-specific header.

## Level 1: PAT Passthrough — Zero Infrastructure

> This is the simplest production-ready configuration. Many enterprise MCP servers — Atlassian, Glean, Sourcegraph — accept per-user authentication via custom HTTP headers carrying a personal access token (PAT). Every engineer already has one. The gateway just needs to deliver it to the right backend and no other:

Reviewer: Small phrasing note: "...deliver it to the right backend and no other:" reads a little clipped. Might read more smoothly as "...deliver it to the right backend, and only that one:" — but leaving to your judgment since the terse style is used elsewhere in the post too.

When a backend expects a different header name than your clients send, rename in transit:

```yaml
forwardHeaders:
  - name: Authorization # extract from the inbound request
    backendHeader: X-User-Token # forward under this name
```

The rename is also how per-user identity and a service-account credential coexist on one backend: forward the user's token under a dedicated header while the backend's `securityPolicy` injects the service credential into `Authorization`. Note that credentials injected by a backend `securityPolicy` are applied at the Envoy layer and overwrite any forwarded header of the same name — so give the forwarded user identity its own header rather than expecting it to take precedence.

Two honest caveats. First, in this mode the gateway is not validating the PATs — the backend is. Keep inbound authentication at the front door (an `MCPRoute` security policy, or OIDC at the gateway) so the gateway itself is never an open proxy. Second, forward identity headers only to backends inside your trust perimeter; if a backend cannot validate the credential, it has no business receiving it.

> ## Level 2: OIDC + JWT Forwarding — the Same-IdP Pattern

Reviewer: ```suggestion
## Level 2: OIDC + JWT Forwarding — The Same-IdP Pattern
```
